File: MineInAbyss/src/main/java/com/derongan/minecraft/mineinabyss/plugin/Ascension/Effect/Effects/SoundAscensionEffect.java
package com.derongan.minecraft.mineinabyss.plugin.Ascension.Effect.Effects;

import com.derongan.minecraft.mineinabyss.plugin.AbyssContext;
import org.bukkit.Location;
import org.bukkit.Sound;
import org.bukkit.entity.Player;
import org.bukkit.util.Vector;

import java.util.List;
import java.util.Objects;
import java.util.Random;

import static java.util.stream.Collectors.toList;


public class SoundAscensionEffect extends AbstractAscensionEffect {
    private List<Sound> sounds;
    private static Random random = new Random();

    private int lastDurationRemaining;
    private int ticksBetweenSoundInitialization;

    public SoundAscensionEffect(AbyssContext context, long offset, int strength, int duration, List<String> allowedSounds) {
        super(context, offset, strength, duration);
        sounds = allowedSounds.stream().map((soundString) -> {
            try {
                return Sound.valueOf(soundString);
            } catch (IllegalArgumentException iae) {
                return null;
            }
        }).filter(Objects::nonNull).collect(toList());
        lastDurationRemaining = durationRemaining;
        ticksBetweenSoundInitialization = strength;
    }

    @Override
    void applyEffect(Player player) {
        if (lastDurationRemaining - durationRemaining >= ticksBetweenSoundInitialization) {
            Location soundLocation = Vector.getRandom().multiply(5).subtract(new Vector(2.5, 2.5, 2.5)).toLocation(player.getWorld()).add(player.getLocation());
            Sound sound = sounds.get(random.nextInt(sounds.size()));

            getContext().getPlugin().getServer().getScheduler().scheduleSyncDelayedTask(getContext().getPlugin(), () -> {
                player.playSound(soundLocation, sound, 1f, 1f);
            }, random.nextInt(getContext().getTickTime()));
            lastDurationRemaining = durationRemaining;
        }
    }
}

File: src/main/java/com/derongan/minecraft/mineinabyss/ascension/effect/effects/SoundAscensionEffect.kt
package com.derongan.minecraft.mineinabyss.ascension.effect.effects

import com.derongan.minecraft.mineinabyss.ascension.effect.AbstractAscensionEffect
import com.derongan.minecraft.mineinabyss.mineInAbyss
import com.mineinabyss.idofront.operators.minus
import com.mineinabyss.idofront.operators.plus
import com.mineinabyss.idofront.operators.times
import com.mineinabyss.idofront.time.TimeSpan
import com.okkero.skedule.schedule
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.bukkit.Sound
import org.bukkit.entity.Player
import org.bukkit.util.Vector
import kotlin.random.Random

@Serializable
@SerialName("sound")
data class SoundAscensionEffect(
    val sounds: List<String>,
    override val offset: TimeSpan,
    override val iterations: Int,
    override val duration: TimeSpan
) : AbstractAscensionEffect() {
    override fun applyEffect(player: Player) { //TODO do not play sounds too quickly together
        val soundLocation = player.location + (Vector.getRandom() * 5 - Vector(2.5, 2.5, 2.5))
        val sound = sounds.random()
        mineInAbyss.schedule {
            waitFor(Random.nextLong(10))
            player.playSound(soundLocation, sound, 1f, 1f)
        }
    }

    override fun clone() = copy()
}
